pkgs/fc/check-ceph/nautilus/fc/check_ceph/check_snapshot_restore.py

```diff
@@ -0,0 +1,229 @@
+    if not len(argv) == 2:
+        print(f"Usage: {argv[0]} <config.toml>")
+        sys.exit(EXIT_CRITICAL)
+
+    try:
+        with open(argv[1], "rt") as configtoml:
+            config = toml.load(configtoml)
+    except (FileNotFoundError, PermissionError):
+        print(f"Error: Unable to open file {argv[1]}.")
+        sys.exit(EXIT_CRITICAL)
+    except toml.TomlDecodeError:
+        print(f"Error: {argv[1]} is not a valid TOML file.")
+        sys.exit(EXIT_CRITICAL)
+
+    # repacking the required values serves as an implicit schema validation
+    try:
+        thresholds = Thresholds(
+            nearfull=float(config["thresholds"]["nearfull"]),
+            full=float(config["thresholds"]["full"]),
+        )
+        pool_roots = config["ceph_roots"]
+        for root, pools in pool_roots.items():
+            assert len(pools) > 0
+            for pool in pools:
+                assert isinstance(pool, str)
+    except (KeyError, ValueError, AssertionError):
+        print(
+            f"Error loading config file {argv[1]}: Values missing or of wrong type."
+        )
+        sys.exit(EXIT_CRITICAL)
+
+    return (thresholds, pool_roots)
+
+
+def eval_fill_warnings(cluster_fill_stats, largest_snaps, thresholds) -> int:
+    """Evaluates the expected fill levels after restoring a cluster and generates the
+    appropriate warnings if necessary.
+    Returns the exit code equivalent to the Sensu warning urgency, calling `sys.exit` is
+    left to the caller."""
+    exit_code = 0
+    if not largest_snaps:
+        print(
+            "INFO: There are no snapshots in the considered pools, nothing to do."
+        )
+        return exit_code
+    for (root, (pool, snap)) in largest_snaps.items():
+        try:
+            root_stats = cluster_fill_stats[root]
+        except KeyError:
+            # FIXME: should this also cause a WARN?
+            print(
+                f"INFO: Root {root} not in fill stats data, something is wrong here."
+            )
+            continue
+        try:
+            fill_now_bytes = int(root_stats["kb_used"] * 1024)
+            total_bytes = int(root_stats["kb"] * 1024)
+        except KeyError:
+            print(f"WARN: Unable to get cluster fill levels for root {root}.")
+            exit_code = max(exit_code, EXIT_WARN)
+        # Note: fill_now_bytes takes into account the sparseness of images and only
+        # includes the actual allocated data in the cluster, while snap["size_bytes"]
+        # describes the provisioned size (100% allocation). This can lead to unintuitive
+        #  (but somehow correct) results.
+        fill_ratio_after_restore = (
+            snap["size_bytes"] + fill_now_bytes
+        ) / total_bytes
+        # Note: This also triggers when the cluster is above the thresholds even without
+        # restoring a anapshot
+        if fill_ratio_after_restore >= thresholds.full:
+            # FIXME: Do we want this to be critical?
```

Comment:
yes but no but yes. we don't want to have to get up at 3am, but we really do want to see this and want to have a ticket.